### DISCUSSION TITLE
Prompt with a command is added to the response.

### DISCUSSION THREAD

**Discussion by ssmyk** [carlmontanari/scrapli#330] [Q&A] [answered]
Sometimes when I read `send_command()` response, it contains the prompt with the executed command as well:

Correct response:
![Zrzut ekranu z 2024-05-06 13-40-32](https://github.com/carlmontanari/scrapli/assets/72394779/5e7835dc-c7b3-4461-9a3b-68d4cb73c38d)
Wrong response:
![Zrzut ekranu z 2024-05-06 13-40-01](https://github.com/carlmontanari/scrapli/assets/72394779/af1d83bb-edfe-4ec4-a9ae-a0dcb13b0e43)

The solution I found is to use `get_prompt()` before sending the command, but it requires being sure that the regex in `comms_prompt_pattern` will catch the prompt. It forces me to check every ssh server prompt if it will be caught, so it would be painful if I want to handle dozens of different servers. Is there any other way to bypass this issue?

**Comment by carlmontanari** [comment]
hmm not sure what you've got going on here -- if the comms prompt pattern is correct/works and you are not disabling strip prompt when doing send command (strip is default iirc) then the prompt should not be in the output. I just quickly tested on a mint box I have sitting around and things work as expected. 

in general, the comms prompt pattern list he most important thing in scrapli and must match the platform(s) you are working with. I would be really really really really really surprised if you had enough variation in the prompts of your platforms that you couldn't get a reasonably readable regex to catch all of them.

what is happening before sending the ping? is there something that is leaving scrapli in a wonky state where it didn't consume the prompt? this really really shouldn't happen because scrpali will not even start writing your command input until after its found (and consumed off the channel/buffer) the prompt. it will then read until it has seen all of your input chars -- again, this consumes this off the channel/buffer. it is possible the device repaints this (but is really unlikely in this case since normal nix things it seems?).

**Reply by ssmyk** [reply]
I tried to reproduce this error on a fresh setup. I created a container from this image: https://registry.hub.docker.com/r/linuxserver/openssh-server/, and used this simple code to connect through ssh:
```
ssh = AsyncGenericDriver(
    host="10.1.1.2",
    port=2222,
    auth_username="linuxserver.io",
    auth_password="password",
    transport="asyncssh",
    auth_strict_key=False,
)

async def run_ssh():
    await ssh.open()
    response = await ssh.send_command("ls -l")
    print(response.result)

asyncio.run(run_ssh())
```
Correct output looks like:
```
total 4
drwxr-xr-x 1 linuxserver.io linuxserver.io  36 May  8 07:32 logs
drwxr-xr-x 1 linuxserver.io linuxserver.io 262 May  8 07:32 ssh_host_keys
-rw-r--r-- 1 linuxserver.io linuxserver.io   4 May  8 07:32 sshd.pid
```
But from time to time prompt with command apears in response:
```
linuxserver-openssh-server-1:~$ ls -l
total 4
drwxr-xr-x 1 linuxserver.io linuxserver.io  36 May  8 07:32 logs
drwxr-xr-x 1 linuxserver.io linuxserver.io 262 May  8 07:32 ssh_host_keys
-rw-r--r-- 1 linuxserver.io linuxserver.io   4 May  8 07:32 sshd.pid
```

**Reply by carlmontanari** [reply]
Nice, thanks for gettin a repro! I’ll take a peak this weekend hopefully!

**Reply by ssmyk** [reply]
A little update which may help:
- sync version (GenericDriver) works correctly, the problem concerns AsyncGenericDriver and asyncssh transport,
- only the first attempt to send a command and receive a response after establishing a connection is affected. Every next try is successful.

**Reply by carlmontanari** [reply]
thanks! that is helpful! planning on checkin tomorrow, will let ya know what I find!

**Reply by carlmontanari** [reply]
ok, for posterity, using this container setup to match what you said I can repro *very rarely* (very annoying right??)

`docker run -e USER_PASSWORD=password -e USER_NAME=linuxserver.io -e PASSWORD_ACCESS=true -p 2222:2222  --hostname="linuxserver-openssh-server-1" linuxserver/openssh-server:latest`

this I think can happen with sync transports too actually -- at least with paramiko/ssh2. I dont think it can happen w/ telnet or system transport, or asynctelnet for that matter -- because all these do "in channel" auth. (update: yep, confirmed can happen w/ paramiko)

the gist I think is that we open the connection and with the generic driver(s) we dont have any "on_open" function -- in the "normal" (for scrapli at least) case we would be acquiring privilege level or disabling paging or whatever in those "core" drivers. this would consume any of the initial login stuff so you would never see this. 

I *think* (though am open to changing my mind?!) that this is a "feature not a bug" -- if you are using a generic platform you could provide an on_open func that simply calls "get_prompt()" to ensure that the initial login stuff has been consumed. 

thoughts?

**Reply by carlmontanari** [reply]
I slept on it and I think that generic driver should have a default on open that just does a single get prompt and nothing else. I think that would entirely solve this and then you could always overwrite it of course. hows that sound?

**Reply by ssmyk** [reply, marked as answer]
So it seems my workaround will be the final solution? ;) For me, it is okay. I will put the get_prompt method to be called on every opening connection, thanks!

**Reply by carlmontanari** [reply]
haha yea basically -- just took me a while to get there lol :D 

> I will put the get_prompt method to be called on every opening connection, thanks!

when I get around to it the next release you shouldn't have to do that since Ill just put that in the "stock" on_open func ill put on the generic driver!

**Reply by carlmontanari** [reply]
hopefully #333 will fix this up for good :D
